### PR TITLE
Updated Convolution/Cross-Correlation

### PR DESCRIPTION
**Pull request by rprechelt**
Made some small changes to cross-correlation and convolution as below:

1) Added in-place versions of xcorr and conv
2) Added versions that allow external memory to be used for computation result
3) Added docstrings describing behaviour.

### CODE REVIEW

File: src/DSP.jl [hunk 1]
```diff
@@ -1,14 +1,23 @@
 ## DSP.jl ##
 
+## === FUNCTIONS == ##
+
 for (T, suff) in ((Float64, "D"), (Float32, ""))
 
+    """
+    Convolution between an input Vector{T} 'X', and a kernel/filter Vector{T} 'K'.
+
+    Returns: Void. Computation result is stored in 'result'.
```
Comment (simonbyrne) [inline src/DSP.jl]: Your code indicates that it returns a result.

Comment (simonbyrne) [inline src/DSP.jl]: Also, the typical julia convention for functions which mutate an object is that they are suffixed with an exclamation mark, and `result` is the first argument, so the signature would be `conv!(result, X, K)`

File: src/DSP.jl [hunk 2]
```diff
@@ -17,25 +26,90 @@ for (T, suff) in ((Float64, "D"), (Float32, ""))
...
+        function conv!(X::Vector{$T}, K::Vector{$T})
+            conv(X, K, X)
+        end
+    end
+
+
+    """
+    Cross-correlation of two Vector{T}'s 'X' and 'Y'.
+
+    Returns: Void. The result of the computation is stored in 'result'
```
Comment (simonbyrne) [inline src/DSP.jl]: Same note again re mutating convention. Also, you might as well return `result`

**Comment by simonbyrne** [comment]
Thanks. Other than the minor points above, it looks good.

**Comment by rprechelt** [comment]
All the requested changes have been made. 

Apologies for missing the couple of `!`; that was an oversight on my part. I wasn't aware of the "mutated argument first" convention, so I will make sure to follow it in the future.

File: src/DSP.jl (hunk 1 above)
Comment (simonbyrne) [inline src/DSP.jl]: The mutating methods should still return the result (even if this was also provided as an argument), as there's no advantage to returning `Void` (my comment above was that you should change the documentation, not the function)

Comment (rprechelt) [inline src/DSP.jl]: My mistake! This has been changed now; sorry for all these small issues that make reviewing a PR so much more arduous.

File: src/DSP.jl [hunk 3]
```diff
@@ -1,41 +1,127 @@
...
+
+    """
+    In-place convolution between an input Vector{T} 'X', and a kernel/filter Vector{T} 'K'.
+
+    Returns: 'X'. 'X' is overwritten with computation result.
+    """
+    @eval begin
+        function conv!(X::Vector{$T}, K::Vector{$T})
+            conv!(X, K, X)
+            return X
```
Comment (simonbyrne) [inline src/DSP.jl]: this line is now redundant

File: src/DSP.jl [hunk 4]
```diff
@@ -1,41 +1,127 @@
...
+
+    Returns: Vector{T} with length = length(X) + length(K) - 1
+    """
     @eval begin
-        function xcorr(X::Array{$T}, K::Array{$T})
-            ksize = length(K)
+        function conv(X::Vector{$T}, K::Vector{$T})
+            result = Array($T, length(X) + length(K) - 1)
+            conv!(result, X, K)
+            return result
```
Comment (simonbyrne) [inline src/DSP.jl]: also redundant line

File: src/DSP.jl [hunk 5]
```diff
@@ -1,41 +1,127 @@
...
+    """
+    Cross-correlation of two Vector{T}'s 'X' and 'Y'.
+
+    Returns: Vector{T} with length(X) + length(Y) - 1
+    """
+    @eval begin
+        function xcorr(X::Vector{$T}, Y::Vector{$T})
+            result = Array($T, (length(X) + length(Y) - 1))
+            xcorr!(result, X, Y)
             return result
```
Comment (simonbyrne) [inline src/DSP.jl]: redundant as well

File: src/DSP.jl [hunk 6]
```diff
@@ -1,41 +1,127 @@
...
+    """
+    In-place cross-correlation of two Vector{T}'s 'X' and 'Y'.
+
+    Returns: 'X'. 'X' is overwritten with the result of the cross-correlation.
+    """
     @eval begin
-        function xcorr(X::Array{$T})
+        function xcorr!(X::Vector{$T}, Y::Vector{$T})
+            xcorr!(X, Y, X)
+            return X
```
Comment (simonbyrne) [inline src/DSP.jl]: and this one.

**Comment by simonbyrne** [comment]
Thanks, and sorry for all the nitpicking, I really do appreciate your help!

**Comment by rprechelt** [comment]
Nitpicking is fantastic! I am just as nitpicky with my own projects so I understand it perfectly... I'm always in the mood to learn new things in languages (even ones that I have been using for years).

This last change should cover the redundant return statements. I was always explicit with my return statements having cut my teeth on C, but I can appreciate the simplicity of the Julia/Matlab convention.

File: src/DSP.jl [hunk 7]
```diff
@@ -1,41 +1,123 @@
...
+
+
+    """
+    In-place convolution between an input Vector{T} 'X', and a kernel/filter Vector{T} 'K'.
+
+    Returns: 'X'. 'X' is overwritten with computation result.
+    """
+    @eval begin
+        function conv!(X::Vector{$T}, K::Vector{$T})
+            conv!(X, K, X)
```
Comment (simonbyrne) [inline src/DSP.jl]: Should this be `conv!(X, X, K)`?

File: src/DSP.jl [hunk 8]
```diff
@@ -1,41 +1,123 @@
...
+
+
+    """
+    In-place cross-correlation of two Vector{T}'s 'X' and 'Y'.
+
+    Returns: 'X'. 'X' is overwritten with the result of the cross-correlation.
+    """
+    @eval begin
+        function xcorr!(X::Vector{$T}, Y::Vector{$T})
+            xcorr!(X, Y, X)
```
Comment (simonbyrne) [inline src/DSP.jl]: And this `xcorr!(X, X, Y)`?

**Comment by rprechelt** [comment]
Ok. This should be all good now. My sincerest apologies for the state of this PR; there's been more issues with this PR than any other PR I've made and I'm sorry you had to be the one to deal with this. Hope it looks good now!

File: src/DSP.jl [hunk 9]
```diff
@@ -1,41 +1,123 @@
 ## DSP.jl ##
 
+## === TYPES === ##
+immutable DFTSetup{T}
+    setup::Ptr{Void}
+
+    direction::Integer
+end
```
Comment (simonbyrne) [inline src/DSP.jl]: I think this belongs to the other PR, does it not?

**Comment by simonbyrne** [comment]
That's fine: these things take time and practice. See my last point, then I think this is good to merge.

**Comment by rprechelt** [comment]
Yep; in all these branches and updates, things were starting to get a little crossed. The change has been made!

**Comment by rprechelt** [comment]
No problem at all; I love doing this stuff!

**Pull request merged**